src/model/project.ts

```diff
@@ -0,0 +1,102 @@
+/**
+ * @license
+ * Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+import {Warning} from '../warning/warning';
+
+import {Document, FeatureKinds} from './document';
+import {Feature} from './feature';
+
+export class Project {
+  private _rootDocuments: Set<Document>;
+  private _toplevelWarnings: Warning[];
+
+  constructor(rootDocuments: Iterable<Document>, warnings: Warning[]) {
+    this._rootDocuments = new Set(rootDocuments);
+    this._toplevelWarnings = warnings;
+
+    // This is a performance optimization. We only need a set of documents such
+    // that all other documents we're interested in can be reached from them.
+    for (const doc of this._rootDocuments) {
+      for (const imprt of doc.getByKind('import')) {
+        // When there's cycles we can keep any element of the cycle, so why not
+        // this one.
+        if (imprt.document !== doc) {
+          this._rootDocuments.delete(imprt.document);
+        }
+      }
+    }
+  }
+
+  getByKind<K extends keyof FeatureKinds>(kind: K): Set<FeatureKinds[K]>;
```

Comment:
I'm cool with that, if just to enforce that they be mutually compatible. What would we call such a thing?